libs/lakebase/connect.go

```diff
@@ -5,16 +5,66 @@ import (
 	"errors"
 	"fmt"
 	"os"
+	"os/exec"
 	"strings"
+	"time"
 
 	"github.com/databricks/cli/libs/cmdctx"
 	"github.com/databricks/cli/libs/cmdio"
-	"github.com/databricks/cli/libs/exec"
+	execlib "github.com/databricks/cli/libs/exec"
 	"github.com/databricks/databricks-sdk-go/service/database"
 	"github.com/google/uuid"
 )
 
+const (
+	// Default retry configuration
+	defaultMaxRetries    = 3
+	defaultInitialDelay  = 1 * time.Second
+	defaultMaxDelay      = 10 * time.Second
+	defaultBackoffFactor = 2.0
+)
+
+// RetryConfig holds configuration for connection retry behavior
+type RetryConfig struct {
+	MaxRetries    int
+	InitialDelay  time.Duration
+	MaxDelay      time.Duration
+	BackoffFactor float64
+}
+
+// tryPsqlInteractive launches psql interactively and returns an error if connection fails
+func tryPsqlInteractive(ctx context.Context, args, env []string) error {
+	cmd := exec.CommandContext(ctx, args[0], args[1:]...)
+	cmd.Env = env
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	err := cmd.Run()
+	if err != nil {
+		// Check if the error might be due to connection issues
+		// Since we can't capture stderr when running interactively, we check the exit code
+		var exitError *exec.ExitError
+		if errors.As(err, &exitError) {
+			exitCode := exitError.ExitCode()
+
+			// We do not use the Databricks SDK for checking whether the error is retryable because the call in question is not to the API
+			// psql returns exit code 2 for connection failures
+			if exitCode == 2 {
+				return fmt.Errorf("connection failed (retryable): psql exited with code %d", exitCode)
+			}
+		}
+		return err
+	}
+
+	return nil
+}
+
 func Connect(ctx context.Context, databaseInstanceName string, extraArgs ...string) error {
+	return ConnectWithRetryConfig(ctx, databaseInstanceName, nil, extraArgs...)
+}
+
+func ConnectWithRetryConfig(ctx context.Context, databaseInstanceName string, retryConfig *RetryConfig, extraArgs ...string) error {
```

Comment:
Given that you have 2 functions:

- Connect() (default retry config)
- ConnectWithRetryConfig (custom retry config)

it is expected that the second one always receives RetryConfig. Thus I'd pass "retryConfig RetryConfig" there (no pointer) and make default object a package level var.